torchtext/experimental/datasets/language_modeling.py

```diff
@@ -65,40 +58,52 @@ def _setup_datasets(dataset_name, tokenizer=None, root='.data', vocab=None,
-    text_transform = sequential_transforms(tokenizer)
 
     if isinstance(data_select, str):
-        data_select = [data_select]
+        data_select = (data_select,)
     if not set(data_select).issubset(set(('train', 'valid', 'test'))):
         raise TypeError('Given data selection {} is not supported!'.format(data_select))
 
     if not single_line and dataset_name != 'WikiText103':
         raise TypeError('single_line must be True except for WikiText103')
-    if dataset_name == 'WMTNewsCrawl':
-        train, = raw.DATASETS[dataset_name](root=root, data_select=('train',))
-        if single_line:
-            raw_data = {'train': [" ".join([txt for txt in train]), ]}
-        else:
-            raw_data = {'train': [txt for txt in train]}
-    else:
-        train, test, valid = raw.DATASETS[dataset_name](root=root, data_select=('train', 'test', 'valid'))
-        # Cache raw text iterable dataset
-        if single_line:
-            raw_data = {'train': [" ".join([txt for txt in train]), ],
-                        'valid': [" ".join(txt for txt in valid), ],
-                        'test': [" ".join(txt for txt in test), ]}
-        else:
-            raw_data = {'train': [txt for txt in train],
-                        'valid': [txt for txt in valid],
-                        'test': [txt for txt in test]}
+
+    # WMTNewsCrawl will throw error if data_select isn't train
+    raw_iter_ = raw.DATASETS[dataset_name](root=root, data_select=data_select)
+    raw_iter = {}
+    for i, name in enumerate(data_select):
+        raw_iter[name] = raw_iter_[i]
 
     if vocab is None:
         if 'train' not in data_select:
             raise TypeError("Must pass a vocab if train is not selected.")
-        vocab = build_vocab(raw_data['train'], text_transform)
-    text_transform = sequential_transforms(text_transform, vocab_func(vocab),
-                                           totensor(dtype=torch.long))
-    return tuple(LanguageModelingDataset(raw_data[item], vocab, text_transform, single_line)
+        # Build vocab from lines of text even if all to be concatenated
+        # for better user experience
+        vocab = build_vocab(raw_iter['train'], tokenizer)
+        # Repopulate with fresh iterator
+        raw_iter['train'] = raw.DATASETS[dataset_name](root=root, data_select='train')
+
+    # Single-line dataset stores numericalized version of dataset. Let's
+    # avoid using extra memory by applying the transforms now instead of later.
+    def text_transform(data, filter_empty=False):
+        for line in data:
+            ids = []
+            for token in tokenizer(line):
+                ids.append(vocab[token])
+            if filter_empty and len(ids) == 0:
+                continue
+            yield torch.tensor(ids, dtype=torch.long)
+
+    raw_data = {}
+    for name in raw_iter:
+        # Materialize datasets
+        raw_data[name] = [torch.tensor(text_transform(txt), dtype=torch.long) for txt in raw_iter[name]]
```

Comment:
Do we decide to call `transform` during `__getitem__` func or within `_setup_datasets` func? For now, the `_setup_datasets` func in the experimental datasets just generate the transforms and we pass transforms and data to the dataset abstraction (for example, `LanguageModelingDataset`).

Reply:
It's inconsistent. For the single_line version the transform is applied during construction (init function of LanguageModelingDataset), else it's applied at runtime during get_item.

Reply:
I think we need a decision for that. With the transformation during the construction, we spend more time on dataset construction, but less time for `__getitem__` (a.k.a. training time). With the opposite, we have a faster construction.

Reply:
This also depends on whether we defined "get_vocab" as a read-only method.